src/GafferScene/Constraint.cpp

```diff
@@ -111,22 +147,358 @@ bool ptInPolygon( const Indexer indexer, const Imath::Vec2< T >& v, const int n
 	return result;
 }
 
-void constructLocalFrame( Imath::M44f& m, const Imath::V3f& p, const Imath::V3f& t, const Imath::V3f& b )
+template< typename PIndexer, typename UVIndexer >
+Imath::V3f interpolateConvexPolygon( const PIndexer points, const UVIndexer uvs, const Imath::V2f& uv, const int n )
 {
-	// ignore tangents with zero length to avoid local frame collapse
+	// NOTE : wachspress coordinates are not defined along edges of polygon
+	//        when uv matches one of more polygon vertices average all corresponding points
+	//        when uv lies on an edge lerp between all points corresponding to end vertices
 
-	const float threshold = 2.f * std::numeric_limits< float >::min();
+	Imath::V3f p( 0.f );
 
-	if(
-		( t.length2() <= threshold ) ||
-		( b.length2() <= threshold ) )
+	float ws = 0.f;
+	for( int i = 0; i < n; ++i )
 	{
-		m.translate( p );
+		if( uv == uvs( i ) )
+		{
+			p += points( i );
+			ws += 1.0f;
+		}
+	}
+
+	if( ws == 0.f )
+	{
+		for( int i = 0; i < n; ++i )
+		{
+			int ip = ( i + n - 1 ) % n;
+			int in = ( i     + 1 ) % n;
+
+			const Imath::V2f vi = uvs( i );
+
+			float Av = ( vi - uvs( ip ) ) % ( uvs( in ) - vi );
+
+			if( Av < 0.f )
+			{
+				// NOTE : reverse the winding order
+
+				Av = -( Av );
+				using std::swap;
+				swap( ip, in );
```

Comment:
in this case there is no benefit, as we know the parameters are ints, but this is the normal way of calling swap. swap can be customised for user types, and is best declared in the same namespace as the user type, which will not be found when called as std::swap( ... ), you need swap( ... ) so that adl kicks in, however that won't find the overloads in namespace std for builtin types, hence the using statement.